gpkit/constraints/model.py

```diff
@@ -38,24 +36,56 @@ class Model(CostedConstraintSet):
     `program` is set during a solve
     `solution` is set at the end of a solve
     """
-    _nums = defaultdict(int)
+
+    # name and num identify a model uniquely
     name = None
     num = None
+    # naming holds the name and num evironment in which a model was created
+    # this includes its own name and num, and those of models containing it
+    naming = None
     program = None
     solution = None
 
-    def __init__(self, cost=None, constraints=None,
-                 substitutions=None, name=None):
+    def __new__(cls, *args, **kwargs):
+        obj = super(Model, cls).__new__(cls, *args, **kwargs)
+        if cls.__name__ != "Model" and not hasattr(cls, "setup"):
+            obj.name = cls.__name__
+            obj.num, obj.naming = begin_variable_naming(obj.name)
+        return obj
+
+    def __init__(self, cost=None, constraints=None, *args, **kwargs):
+        unused_vars = None
+        substitutions = kwargs.pop("substitutions", None)  # reserved keyword
+        if hasattr(self, "setup"):
+            with NamedVariables(self.__class__.__name__):
+                start_args = [cost, constraints]
+                args = tuple(a for a in start_args if a is not None) + args
```

Comment:
It adds the `*args` from the function parameters on the end, after the unused cost, constraints.